proposals/flexible-vectors/README.md

```diff
@@ -1,7 +1,1184 @@
+    return result
+```
+
+### Mask last
+
+Returns a mask with a single active lane that corresponds to the last active lane of the input.
+If there is no active lanes, then an empty mask is returned.
+
+- `vec.m8.last(m: vec.m8) -> vec.m8`
+- `vec.m16.last(m: vec.m16) -> vec.m16`
+- `vec.m32.last(m: vec.m32) -> vec.m32`
+- `vec.m64.last(m: vec.m64) -> vec.m64`
+- `vec.m128.last(m: vec.m128) -> vec.m128`
+
+```python
+def vec.S.last(m):
+    result = mask.S.New()
+    idx = -1
+    for i in range(vec.S.length):
+        if m[i]:
+            idx = i
+    if idx >= 0:
+        result[idx] = 1
+    return result
+```
+
+
+## Memory operations
+
+### Vector load
+
+- `vec.v8.load(a: memarg) -> vec.v8`
+- `vec.v16.load(a: memarg) -> vec.v16`
+- `vec.v32.load(a: memarg) -> vec.v32`
+- `vec.v64.load(a: memarg) -> vec.v64`
+- `vec.v128.load(a: memarg) -> vec.v128`
+
+### Vector load mask zero
+
+Inactive elements are set to `0`
+
+- `vec.v8.load_mz(m: vec.m8, a: memarg) -> vec.v8`
+- `vec.v16.load_mz(m: vec.m16, a: memarg) -> vec.v16`
+- `vec.v32.load_mz(m: vec.m32, a: memarg) -> vec.v32`
+- `vec.v64.load_mz(m: vec.m64, a: memarg) -> vec.v64`
+- `vec.v128.load_mz(m: vec.m128, a: memarg) -> vec.v128`
+
+### Vector load mask undefined
+
+Inactive elements have undefined values
+
+- `vec.v8.load_mx(m: vec.m8, a: memarg) -> vec.v8`
+- `vec.v16.load_mx(m: vec.m16, a: memarg) -> vec.v16`
+- `vec.v32.load_mx(m: vec.m32, a: memarg) -> vec.v32`
+- `vec.v64.load_mx(m: vec.m64, a: memarg) -> vec.v64`
+- `vec.v128.load_mx(m: vec.m128, a: memarg) -> vec.v128`
+
+### Vector load splat
+
+- `vec.v8.load_splat(a: memarg) -> vec.v8`
+- `vec.v16.load_splat(a: memarg) -> vec.v16`
+- `vec.v32.load_splat(a: memarg) -> vec.v32`
+- `vec.v64.load_splat(a: memarg) -> vec.v64`
+- `vec.v128.load_splat(a: memarg) -> vec.v128`
+
+### Vector store
+
+- `vec.v8.store(a: memarg, v: vec.v8)`
+- `vec.v16.store(a: memarg, v: vec.v16)`
+- `vec.v32.store(a: memarg, v: vec.v32)`
+- `vec.v64.store(a: memarg, v: vec.v64)`
+- `vec.v128.store(a: memarg, v: vec.v128)`
+
+### Vector store mask
+
+Inactive elements are not stored
+
+- `vec.v8.m_store(m: vec.m8, a: memarg, v: vec.v8)`
+- `vec.v16.m_store(m: vec.m16, a: memarg, v: vec.v16)`
+- `vec.v32.m_store(m: vec.m32, a: memarg, v: vec.v32)`
+- `vec.v64.m_store(m: vec.m64, a: memarg, v: vec.v64)`
+- `vec.v128.m_store(m: vec.m128, a: memarg, v: vec.v128)`
+
+## Lane operations
+
+### Splat scalar
+
+`idx` is interpreted modulo the length of the vector.
+
+- `vec.i8.splat(v: vec.v8, x: i32) -> vec.v8`
+- `vec.i16.splat(v: vec.v16, x: i32) -> vec.v16`
+- `vec.i32.splat(v: vec.v32, x: i32) -> vec.v32`
+- `vec.f32.splat(v: vec.v32, x: f32) -> vec.32`
+- `vec.i64.splat(v: vec.v64, x: i64) -> vec.v64`
+- `vec.f64.splat(v: vec.v64, x: f64) -> vec.v64`
+- `vec.v128.splat(v: vec.v128, x: v128) -> vec.v128`
+
+
+### Extract lane
+
+`idx` is interpreted modulo the length of the vector.
+
+- `vec.s8.extract_lane(v: vec.v8, idx: i32) -> i32`
```

Comment:
I do expect indices to be almost always constants. And I do expect WASM engines to detect such constants with a constant folding pass (which should be both simpler and more robust than pattern matching, I assume).

We could have two instructions: one with runtime indices, and one with immediates. But I see two problems:
- it doubles the number of opcodes,
- except for 0, the index will usually not be an immediate as it will most likely depends on the vector length.

The vector length is a "platform constant", so can be detected by the engine as being constant, but is not an immediate. So constant detection looks more useful than immediates.

Reply:
>     * except for 0, the index will usually not be an immediate as it will most likely depends on the vector length.

BTW that might apply to more values, depending on how the discussion about the permissible vector lengths goes (given that you seem to be leaning towards powers of 2). E.g. in the case of SVE everything that fits within the lowest 16 bytes is fair game.